### ISSUE TITLE
Proposal: custom observable operator for inflating relationships

### ISSUE THREAD

**Issue by ericeslinger** [kstf/plump#93]
I've been thinking about how to handle sort order ( #88 ) and nested observables. It is [pretty straightforward](https://github.com/ReactiveX/rxjs/blob/master/doc/operator-creation.md) to create a new operator and subclass Observable to have that operator. So, what if we had

```
Model.asObservable() => PlumpObservable<ModelData>
class PlumpObservable extends Observable {
  inflateRelationship(relName: string) => PlumpObservable<(ModelData & {meta: any})[]>;
}
```

So then we have a hook into PlumpObservable (all calls to `Model.asObservable` return our modified Observable, and all chained calls like `.flatMap` and whatever continue to return the PlumpObservable type, if you override `Observable.lift` in the PlumpObservable class, according to the linked documentation.

This operator would, when acting on an `Observable<ModelData>`:
* subscribe to the Observable
* take the `ModelData.relationships[relName]` and convert it into an array of Models
* do an Observable.combineLatest on each of those Model.asObservables
* of course manage internal subscription state to handle unsubscribes when the array changes

So the resultant observable would be `Observable<ModelData[]>` that would fire an update whenever either the array changed (add / delete values) or the child ModelData changed (name change in a profile on a participants list).

Then, when encountering the typical "I need the names of the editors, in alphabetical order" situtation on a document, we could do:

```
const editorsSortedByName = 
  doc
  .asObservable('relationships.participants') 
  // the base ModelData for the document
  .inflateRelationship('relationships.participants') 
  // an observable of arrays of ModelData<Profile> types
  .map(profilesData => profilesData.filter(i => i.meta.perm >= 3))
  .map(editors => editors.sort(
    (a, b) => a.attributes.short_text.localCompare(b.attributes.short_text)
  );
// NOTE: null protection omitted for clarity here.
```

I'm trying to figure out the minimal tool that gives us large boiler-plate reduction for most cases, and rather than muck around with a DSL on the string argument to `asObservable()`, it feels like just adding some sugar as an operator is better.

**Comment by ericeslinger** [comment]
@mrpinsky what do you think about this? Would it handle some / most / all of our use cases? I think it's pretty straightforward, but I don't want to spend time working on it if it isn't the right approach.

**Comment by mrpinsky** [comment]
@ericeslinger This seems similar to an approach I was considering on an ad hoc basis. It seems to cover our two main use cases of sorting and filtering an Observable array of Observables. The main thing that prevented me from pursuing it in the past was the complexity of the `combineLatest` step. Since `Observable.combineLatest` only works on 2 observables at a time, we would need to do something like
```Array.reduce((acc, curr) => acc.combineLatest(curr, (list, data) => list.concat(data)))```
which I was wary of because I didn't (and still don't) know whether that would make managing subscriptions on the individual Observables difficult. If that is a manageable issue, I say go for it.

**Comment by ericeslinger** [comment]
I'm pretty sure I can use combinelatest on an array, if I use spread operators (like `Observable.combineLatest(...arr).map((...arr) => arr[1])` sort of thing. But that's just a hypothesis based on reading the documentation, rather than really digging in. If it seems like a reasonable thing to have working though, I'll play with it soon.

**Comment by ericeslinger** [comment]
woops, here's the hard part. I was fiddling with this because I needed to change some more of the API surface on plump (so the login system can get access to plump-store-rest's socket.io instance when necessary) and was going to put this in too.

when I'm doing the inflate, I need to do plump.find() on all the children, but in that particular closure I don't actually have a plump handy; I just have the ModelData that is coming up the observable stream.

I think I can hack it in to the observable interface (basically stick a `(public plump: Plump)` argument into the PlumpObservable constructor, since I control when the constructor is being called at all times anyway. Feels a bit hacky, but it's not the end of the world.

Mostly noting here so I remember why I did this when I see this code in a year.
